cudaPackages: enable cross-compilation (take two)

ConnorBaker opened the pull request:
> [!Important]
>
> This PR includes changes which have not yet been merged into `master`. It should not be merged prior to them:
>
> - #296219
>   - NOTE: This PR targeted `staging` and must be merged into `master` before merging.
> - #281371

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

#275560 brings in a bunch of additional work that I thought needed to be coupled with support for cross-compilation, but over the break I decided to revisit this. Maybe it doesn't need to be coupled -- this is a second attempt at enabled cross-compilation for cudaPackages.

Of note, both `cuda-modules/setup-hooks/extension.nix` and `cuda-modules/cuda/overrides.nix` have been refactored. They are now attribute sets of functions which are invoked with `callPackage` to create setup hooks and package overrides, respectively. The move to using `final.callPackage` rather than retrieving required packages directly from `final` allows us to use the `__spliced` attribute on derivations when cross-compiling to ensure we're selecting the correct version of a package.

When adding packages to `nativeBuildInputs` or `buildInputs`, splicing (largely) happens automatically. That is, there is no need to manually specify which splice a package should be drawn from. Additionally, the default package set the `callPackage` arguments are drawn from, `pkgs` (aka `pkgsHostTarget`), helps minimize breakage. However, there are places where we *do* need to specify the splice to use:

- Using specific package outputs: splices must be manually inserted prior to using a particular output of a package, as otherwise the output will be drawn from the splice corresponding to `pkgs` (which is to say `pkgsHostTarget`).
- Phases: any usage of a package supplied by `nativeBuildInputs` in `pre/post` phases should be spliced so the correct package is made available.
- Flags: any usage of a package supplied by `nativeBuildInputs` in a `*Flags`-style argument should be spliced, again to ensure the correct package is made available and that the default (`pkgs`) splice isn't used instead.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
cudaPackages: cross-compilation attempt 2

workaround bug in cross-compilation / meta.getExe

cudaPackages.autoAddCudaCompatRunpathHook: correct meta.platforms reference

cudaPackages.backendStdenv: use coreutils from buildPackages when cross-compiling

ConnorBaker commented:
Unsure why `buildPlatform` and `hostPlatform` as seen in `cudaPackages.flags` (in the `isJetsonBuild` check) are both `x86_64-linux`, while in `backendStdenv` they're correctly marked as `x86_64-linux` and `aarch64-linux` respectively when trying to build things with `legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform.cudaPackages.*`.

For reference, using the following Nixpkgs config on my `x86_64-linux` desktop when trying to cross-compile for Jetson:

```nix
{
  allowUnfree = true;
  allowBroken = false;
  cudaSupport = true;
  cudaCapabilities = ["8.7"];
}
```

yannham commented:
It's been discussed elsewhere that PR #275947 could help with current build issues (nvcc fails because `autopatchElf` is unable to find libstdc++). I tried to merge the two and cross-compile nvcc for aarch64 on an x86_64 machine
 (`.#legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform.cudaPackages.cuda_nvcc`): it doesn't fail on autopatchElf, but it doesn't build either, see attached logs.

For the record, I'm able to cross-compile NVCC on this PR with the following (stupid monkey)-patch:

```diff
diff --git a/pkgs/build-support/setup-hooks/auto-patchelf.sh b/pkgs/build-support/setup-hooks/auto-patchelf.sh
index 371389df427b..53d732432ecd 100644
--- a/pkgs/build-support/setup-hooks/auto-patchelf.sh
+++ b/pkgs/build-support/setup-hooks/auto-patchelf.sh
@@ -5,6 +5,7 @@ declare -a extraAutoPatchelfLibs
 
 gatherLibraries() {
     autoPatchelfLibs+=("$1/lib")
+    autoPatchelfLibs+=("$1/aarch64-unknown-linux-gnu/lib")
 }
 
 # shellcheck disable=SC2154
```

This is of course ad-hoc, and the issue is to find a clean way to get the same effect.

[gcc-aarch64.log](https://github.com/NixOS/nixpkgs/files/13902099/gcc-aarch64.log)

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
@yannham the idea was to replace https://github.com/NixOS/nixpkgs/blob/e371d547840cf69e85f2a50356e94823b4a71b15/pkgs/development/cuda-modules/generic-builders/manifest.nix#L197

with `backentStdenv.cc.cxxStdlib.solib` (didn't check the path)

SomeoneSerge replied:
This would be a string rather than a derivation, and it'd be a store path with the `aarch64-blahblah/lib` suffix

SomeoneSerge replied:
OTOH this looks like a bug in autopatchelf that it doesn't know where to look for the libraries in the cross compilation setting

yannham replied:
> OTOH this looks like a bug in autopatchelf that it doesn't know where to look for the libraries in the cross compilation setting

I agree. This doesn't seem to be really related to CUDA, but could theoretically bite other packages.

SomeoneSerge replied:
https://github.com/NixOS/nixpkgs/pull/282236 appears to be a likely longer-term solution, but we can't rely on it until it went through staging

yannham commented:
@SomeoneSerge with the following patch to the combination of sthis PR and https://github.com/NixOS/nixpkgs/pull/275947:

```diff
--- a/pkgs/development/cuda-modules/generic-builders/manifest.nix
+++ b/pkgs/development/cuda-modules/generic-builders/manifest.nix
@@ -194,7 +194,7 @@ backendStdenv.mkDerivation (
         # one that is compatible with the rest of nixpkgs, even when
         # nvcc forces us to use an older gcc
         # NB: We don't actually know if this is the right thing to do
-        stdenv.cc.cc.lib
+        backendStdenv.cc.cxxStdlib.solib
       ];
 
     # Picked up by autoPatchelf
```

I get the following build error, which looks very similar to trying to build without the above diff the merge of the two PRs:

[snd.log](https://github.com/NixOS/nixpkgs/files/13919641/snd.log)

SomeoneSerge commented:
H'mmm, maybe we got to keep the original output for the includes to be discoverable by the hook. What if you keep both `backendStdenv.cc.cxxStdlib.package` and `backendStdenv.cc.cxxStdlib.solib`?

ConnorBaker commented:
@yannham I just rebased and force-pushed. Would you mind taking a look at this? Looks like `autoPatchelf` isn't correctly ignoring things specified as overrides. Any ideas if it's because the `autoPatchelf` setup hook is coming from a different instance of `nixpkgs` (`buildPackages` I think since it needs to run on the machine hosting the build during the build)?

Target a Jetson build (cross-compiling from `x86_64-linux` to `aarch64-linux`)

```console
$ cat ~/.config/nixpkgs/config.nix 
{
  allowUnfree = true;
  allowBroken = false;
  cudaSupport = true;
  cudaCapabilities = ["8.7"];
}
```

with

```console
$ nom build --impure .#legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform.cudaPackages.cuda_compat
warning: Git tree '/home/connorbaker/nixpkgs' is dirty
this derivation will be built:
  /nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv
cuda_compat-aarch64-unknown-linux-gnu> building '/nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv' on 'ssh-ng://nix@nixos-build01'
copying 0 paths
cuda_compat-aarch64-unknown-linux-gnu> building '/nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv'
cuda_compat-aarch64-unknown-linux-gnu> structuredAttrs is enabled
cuda_compat-aarch64-unknown-linux-gnu> Sourcing auto-add-opengl-runpath-hook
cuda_compat-aarch64-unknown-linux-gnu> Using autoAddOpenGLRunpathPhase
cuda_compat-aarch64-unknown-linux-gnu> Sourcing mark-for-cudatoolkit-root-hook
cuda_compat-aarch64-unknown-linux-gnu> Running phase: unpackPhase
cuda_compat-aarch64-unknown-linux-gnu> unpacking source archive /nix/store/16c7v9bpq3pbh4436vdawvm34bl6knfy-cuda_compat-linux-aarch64-11.8.31339915-archive.tar.xz
cuda_compat-aarch64-unknown-linux-gnu> source root is cuda_compat-linux-aarch64-11.8.31339915-archive
cuda_compat-aarch64-unknown-linux-gnu> setting SOURCE_DATE_EPOCH to timestamp 1661581251 of file cuda_compat-linux-aarch64-11.8.31339915-archive/LICENSE
cuda_compat-aarch64-unknown-linux-gnu> Running phase: patchPhase
cuda_compat-aarch64-unknown-linux-gnu> Running phase: updateAutotoolsGnuConfigScriptsPhase
cuda_compat-aarch64-unknown-linux-gnu> Running phase: updateAutotoolsGnuConfigScriptsPhase
cuda_compat-aarch64-unknown-linux-gnu> Running phase: configurePhase
cuda_compat-aarch64-unknown-linux-gnu> no configure script, doing nothing
cuda_compat-aarch64-unknown-linux-gnu> Running phase: installPhase
cuda_compat-aarch64-unknown-linux-gnu> Running phase: fixupPhase
cuda_compat-aarch64-unknown-linux-gnu> shrinking RPATHs of ELF executables and libraries in /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915
cuda_compat-aarch64-unknown-linux-gnu> shrinking /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libnvidia-ptxjitcompiler.so
cuda_compat-aarch64-unknown-linux-gnu> shrinking /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libnvidia-nvvm.so
cuda_compat-aarch64-unknown-linux-gnu> shrinking /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
cuda_compat-aarch64-unknown-linux-gnu> checking for references to /build/ in /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915...
cuda_compat-aarch64-unknown-linux-gnu> patching script interpreter paths in /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915
cuda_compat-aarch64-unknown-linux-gnu> automatically fixing dependencies for ELF files
cuda_compat-aarch64-unknown-linux-gnu> {'append_rpaths': [PosixPath('$ORIGIN')],
cuda_compat-aarch64-unknown-linux-gnu>  'extra_args': [],
cuda_compat-aarch64-unknown-linux-gnu>  'ignore_missing': ['libcuda.so.1.1 libnvrm_gpu.so libnvrm_mem.so '
cuda_compat-aarch64-unknown-linux-gnu>                     'libnvdla_runtime.so'],
cuda_compat-aarch64-unknown-linux-gnu>  'libs': [PosixPath('/nix/store/yzflgl9pbqkmj54f5xqsxj3xy4rqi2mr-aarch64-unknown-linux-gnu-gcc-13.2.0-lib/lib')],
cuda_compat-aarch64-unknown-linux-gnu>  'paths': [PosixPath('/nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915')],
cuda_compat-aarch64-unknown-linux-gnu>  'recursive': True,
cuda_compat-aarch64-unknown-linux-gnu>  'runtime_dependencies': []}
cuda_compat-aarch64-unknown-linux-gnu> searching for dependencies of /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libnvidia-ptxjitcompiler.so
cuda_compat-aarch64-unknown-linux-gnu> setting RPATH to: $ORIGIN
cuda_compat-aarch64-unknown-linux-gnu> searching for dependencies of /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libnvidia-nvvm.so
cuda_compat-aarch64-unknown-linux-gnu> setting RPATH to: $ORIGIN
cuda_compat-aarch64-unknown-linux-gnu> searching for dependencies of /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
cuda_compat-aarch64-unknown-linux-gnu>     libnvrm_gpu.so -> not found!
cuda_compat-aarch64-unknown-linux-gnu>     libnvrm_mem.so -> not found!
cuda_compat-aarch64-unknown-linux-gnu> setting RPATH to: $ORIGIN
cuda_compat-aarch64-unknown-linux-gnu> auto-patchelf: 2 dependencies could not be satisfied
cuda_compat-aarch64-unknown-linux-gnu> error: auto-patchelf could not satisfy dependency libnvrm_gpu.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
cuda_compat-aarch64-unknown-linux-gnu> error: auto-patchelf could not satisfy dependency libnvrm_mem.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
cuda_compat-aarch64-unknown-linux-gnu> auto-patchelf failed to find all the required dependencies.
cuda_compat-aarch64-unknown-linux-gnu> Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
cuda_compat-aarch64-unknown-linux-gnu> /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 74: pop_var_context: head of shell_variables not a function context
cuda_compat-aarch64-unknown-linux-gnu> /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1454: pop_var_context: head of shell_variables not a function context
cuda_compat-aarch64-unknown-linux-gnu> /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1557: pop_var_context: head of shell_variables not a function context
error: build of '/nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv' on 'ssh-ng://nix@nixos-build01' failed: builder for '/nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv' failed with exit code 1;
       last 10 log lines:
       >     libnvrm_mem.so -> not found!
       > setting RPATH to: $ORIGIN
       > auto-patchelf: 2 dependencies could not be satisfied
       > error: auto-patchelf could not satisfy dependency libnvrm_gpu.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
       > error: auto-patchelf could not satisfy dependency libnvrm_mem.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915/compat/libcuda.so.1.1
       > auto-patchelf failed to find all the required dependencies.
       > Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
       > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 74: pop_var_context: head of shell_variables not a function context
       > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1454: pop_var_context: head of shell_variables not a function context
       > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1557: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv'.
┏━ 1 Errors: 
┃ error: build of '/nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.drv' on 'ssh-ng://n…
┃        last 10 log lines:
┃        >     libnvrm_mem.so -> not found!
┃        > setting RPATH to: $ORIGIN
┃        > auto-patchelf: 2 dependencies could not be satisfied
┃        > error: auto-patchelf could not satisfy dependency libnvrm_gpu.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda…
┃        > error: auto-patchelf could not satisfy dependency libnvrm_mem.so wanted by /nix/store/ikw387y48r90fbfnfvvqr9pzk8x0nbw0-cuda…
┃        > auto-patchelf failed to find all the required dependencies.
┃        > Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
┃        > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 74: pop_var_context: head of shell_variables not a fun…
┃        > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1454: pop_var_context: head of shell_variables not a f…
┃        > /nix/store/y81zm0dylgal3f1v0hk65ld1qn2n56mv-stdenv-linux/setup: line 1557: pop_var_context: head of shell_variables not a f…
┃        For full logs, run 'nix log /nix/store/3mkllbijmvh7dk7xv0qx20ms36hdihv2-cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915.d…
┣━ Dependency Graph:
┃ ⚠ cuda_compat-aarch64-unknown-linux-gnu-11.8.31339915 failed with exit code 1 after ⏱ 0s in fixupPhase
┣━━━                                                     
┗━ ∑ ⚠ Exited after 1 build failures at 15:47:14 after 3s
```

ConnorBaker commented:
A related issue where `coreutils` weren't properly handled by `wrapCCWith` (problematic for us because we rely on `stdenvAdapters.useLibsFrom`, which uses it internally) is handled in #281371.

Commit:
gcc: link $lib/lib -> $lib/$targetConfig correctly and consistently

When native-compiling, gcc will install libraries into:

  /nix/store/...-$targetConfig-gcc-$version-lib/lib

When cross-compiling, gcc will install libraries into:

  /nix/store/...-$targetConfig-gcc-$version-lib/$targetConfig

When cross-compiling, we intended to create a link from $lib/lib to
$lib/$targetConfig, so that downstream users can always safely
assume that "${lib.getLib stdenv.cc.cc}/lib" is where the gcc
libraries are, regardless of whether `stdenv.cc.cc` is a cross
compiler or a native compiler.

Unfortunately, there were two problems with how we were trying to
create these links:

1. The link would be created only when `enableLibGccOutput==true`

2. The link was being created from the incorrect source
   `$lib/lib/lib` instead of `$lib/lib`.

Both of these mistakes are my fault.  This commit corrects them by
creating the link using `ln -Ts` (which is more predictable) and by
creating the link from `gcc/common/builder.nix` rather than from
`gcc/common/libgcc.nix`.

Commit:
cuda-modules/setup-hooks: introduce helper function and add comments about callPackage

Commit:
cuda-modules/generic-builders/manifest: use hostTarget autoAddCudaCompatRunpathHook

Commit:
cuda-modules/cuda/overrides: introduce helper function and add comments about callPackage

Commit:
cuda-modules/cuda/overrides: add TODOs for @connorbaker

ConnorBaker commented on pkgs/development/cuda-modules/cuda/overrides.nix:
```diff
@@ -1,279 +1,400 @@
+    cuda_nvcc.overrideAttrs (
+      prevAttrs: {
+        # Remove once cuda-find-redist-features has a special case for libcuda
+        # TODO(@connorbaker): The order of build outputs matters as we traverse them when creating split outputs.
+        # The `lib` output cannot come after `static` as it moves all the static libraries back to the `lib` output.
+        outputs =
+          prevAttrs.outputs
+          ++ lib.lists.optionals (!(builtins.elem "lib" prevAttrs.outputs)) [ "lib" ];
```
TODO @ConnorBaker

ConnorBaker commented on pkgs/development/cuda-modules/cuda/overrides.nix:
```diff
@@ -1,279 +1,400 @@
-            --replace \
-              '$(TOP)/$(_TARGET_DIR_)/include' \
-              "''${!outputDev}/include"
+        # TODO(@connorbaker): We should specify the spliced version of backendStdenv and cuda_cudart to use here.
```
TODO @ConnorBaker

ConnorBaker commented on pkgs/development/cuda-modules/setup-hooks/extension.nix:
```diff
@@ -1,74 +1,74 @@
-      )
-      {}
-    );
+        substitutions = {
+          # Required in addition to ccRoot as otherwise bin/gcc is looked up
+          # when building CMakeCUDACompilerId.cu
+          ccFullPath = "${backendStdenv.cc}/bin/${backendStdenv.cc.targetPrefix}c++";
+          # Point NVCC at a compatible compiler
+          ccRoot = "${backendStdenv.cc}";
+          setupCudaHook = placeholder "out";
+        };
```
TODO @ConnorBaker

Are these correct? Do we need to use a spliced version? Unclear whether `makeSetupHook` does anything on the backend to draw these from `buildPackages` (`pkgsBuildHost`) or uses the default `pkgs` (`pkgsHostTarget`).

SomeoneSerge replied:
We should use the current host's compiler. Then we take the hook from the BuildHost slice, so its "host" is really our "build"

ConnorBaker commented:
@yannham current blockers are these:

Building `cuda_nvcc` will fail because the order of the outputs matters, and `lib` should not come after `static` as it will move all the libraries out of the `static` output and into the `lib` output, causing an empty `static` output and a build failure: https://github.com/NixOS/nixpkgs/pull/279952#discussion_r1462692606

Fixing that error (either by commenting it out temporarily or inserting `lib` in the correct position) leads to a new build error when trying to compile `legacyPackages.x86_64-linux.pkgsCross.aarch64-multiplatform.cudaPackages.saxpy` -- `cuda_nvcc` from the Jetson (`linux-aarch64` is the NVIDIA redist arch name) tarball being built on `x86_64-linux` fails to link against the `libgcc` and `libc++` libraries from the `buildPackages` compiler (as expected). However, I have no idea why it's trying to do that.

Unclear whether we're sourcing compilers from the correct package sets in https://github.com/NixOS/nixpkgs/pull/279952#discussion_r1462692708 and https://github.com/NixOS/nixpkgs/pull/279952#discussion_r1462693479.

SomeoneSerge commented:
> the order of the outputs matters, and lib should not come after static as it will move all the libraries out of the static output and into the lib output

That's a regression after the cuda12 fixes then, and we should PR a fix separately. I thought we'd just update the manifest generator and remove these `outputs = ... ++ ...` hacks

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -46,7 +46,7 @@ let
   # The redistArch is the name of the architecture for which the redistributable is built.
   # It is `"unsupported"` if the redistributable is not supported on the target platform.
-  redistArch = flags.getRedistArch hostPlatform.system;
+  redistArch = flags.getRedistArch targetPlatform.system;
```
Why? I think we generally want the host system's binaries, except when we handle `lib/` and `nvvm/lib/` in `nvcc` (then we do want target)

SomeoneSerge commented on pkgs/development/cuda-modules/setup-hooks/extension.nix:
```diff
@@ -1,74 +1,74 @@
+    in
+    mapAttrs (_: value: callPackage value { }) setupHooksAttrs;
+in
+createSetupHooks {
```
Not sure I get this. Why do we need `mapAttrs`?

ConnorBaker replied:
`mapAttrs` is used here to apply `callPackage` to the value of each attribute in the attribute set -- it saves us (well... me) from having to wrap the value of each attribute in a call to `callPackage ... {}` (which would increase indentation when using nixfmt).

ConnorBaker replied:
Sure, I can do that!

Commit:
cuda-modules: use hostPlatform when downloading binaries

ConnorBaker commented:
Oh joy, looks like `cc-wrapper` (by way of `stdenvAdapters.useLibsFrom`) might not be creating some symlinks we need.


```console
[connorbaker@nixos-desktop:~/nixpkgs]$ ls "$(nix eval --raw --impure .#pkgsCross.aarch64-multiplatform.gcc12Stdenv.cc.out.outPath)/bin"
warning: Git tree '/home/connorbaker/nixpkgs' is dirty
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
aarch64-unknown-linux-gnu-addr2line  aarch64-unknown-linux-gnu-c++filt  aarch64-unknown-linux-gnu-gcc      aarch64-unknown-linux-gnu-nm       aarch64-unknown-linux-gnu-size
aarch64-unknown-linux-gnu-ar         aarch64-unknown-linux-gnu-cpp      aarch64-unknown-linux-gnu-gprof    aarch64-unknown-linux-gnu-objcopy  aarch64-unknown-linux-gnu-strings
aarch64-unknown-linux-gnu-as         aarch64-unknown-linux-gnu-dwp      aarch64-unknown-linux-gnu-ld       aarch64-unknown-linux-gnu-objdump  aarch64-unknown-linux-gnu-strip
aarch64-unknown-linux-gnu-c++        aarch64-unknown-linux-gnu-elfedit  aarch64-unknown-linux-gnu-ld.bfd   aarch64-unknown-linux-gnu-ranlib
aarch64-unknown-linux-gnu-cc         aarch64-unknown-linux-gnu-g++      aarch64-unknown-linux-gnu-ld.gold  aarch64-unknown-linux-gnu-readelf

[connorbaker@nixos-desktop:~/nixpkgs]$ ls "$(nix eval --raw --impure .#pkgsCross.aarch64-multiplatform.cudaPackages.backendStdenv.cc.out.outPath)/bin"
warning: Git tree '/home/connorbaker/nixpkgs' is dirty
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
addr2line  ar  as  c++filt  dwp  elfedit  gprof  ld  ld.bfd  ld.gold  nm  objcopy  objdump  ranlib  readelf  size  strings  strip
```

ConnorBaker commented:
> Oh joy, looks like `cc-wrapper` (by way of `stdenvAdapters.useLibsFrom`) might not be creating some symlinks we need.
> 
> ```
> [connorbaker@nixos-desktop:~/nixpkgs]$ ls "$(nix eval --raw --impure .#pkgsCross.aarch64-multiplatform.gcc12Stdenv.cc.out.outPath)/bin"
> warning: Git tree '/home/connorbaker/nixpkgs' is dirty
> warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
> aarch64-unknown-linux-gnu-addr2line  aarch64-unknown-linux-gnu-c++filt  aarch64-unknown-linux-gnu-gcc      aarch64-unknown-linux-gnu-nm       aarch64-unknown-linux-gnu-size
> aarch64-unknown-linux-gnu-ar         aarch64-unknown-linux-gnu-cpp      aarch64-unknown-linux-gnu-gprof    aarch64-unknown-linux-gnu-objcopy  aarch64-unknown-linux-gnu-strings
> aarch64-unknown-linux-gnu-as         aarch64-unknown-linux-gnu-dwp      aarch64-unknown-linux-gnu-ld       aarch64-unknown-linux-gnu-objdump  aarch64-unknown-linux-gnu-strip
> aarch64-unknown-linux-gnu-c++        aarch64-unknown-linux-gnu-elfedit  aarch64-unknown-linux-gnu-ld.bfd   aarch64-unknown-linux-gnu-ranlib
> aarch64-unknown-linux-gnu-cc         aarch64-unknown-linux-gnu-g++      aarch64-unknown-linux-gnu-ld.gold  aarch64-unknown-linux-gnu-readelf
> 
> [connorbaker@nixos-desktop:~/nixpkgs]$ ls "$(nix eval --raw --impure .#pkgsCross.aarch64-multiplatform.cudaPackages.backendStdenv.cc.out.outPath)/bin"
> warning: Git tree '/home/connorbaker/nixpkgs' is dirty
> warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
> addr2line  ar  as  c++filt  dwp  elfedit  gprof  ld  ld.bfd  ld.gold  nm  objcopy  objdump  ranlib  readelf  size  strings  strip
> ```

With the most recent changes in #281371, the outputs match.

ConnorBaker commented:
TODO(@ConnorBaker): NVCC, as well as any dependency which finds itself in `nativeBuildInputs`, cannot have a dependency on `cuda_compat` (usually introduced by way of one of our setup hooks) when we are cross-compiling.

Likely for ease of implementation, NVCC is used as a way to inject dependencies on setup hooks. If that's the case, it falls apart when cross-compiling, as our setup hooks need to be able to run on the build platform rather than the host/target.

SomeoneSerge commented:
> Likely for ease of implementation, NVCC is used as a way to inject dependencies on setup hooks. If that's the case, it falls apart when cross-compiling, as our setup hooks need to be able to run on the build platform rather than the host/target.

This just has to do with the fact that `cuda_nvcc` (if it's ever to work) should actually depend on two `src`s: one  for the host platform (`bin/nvcc`, `cicc`, `ptxas`, etc) and one from the target platform (`.a`, `.so`). 

Since we'll always be taking `nvcc` and `setupCudaHook` from the `pkgsBuildHost` slice, it'll have the executables for the build platform and the libraries for the host platform

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -195,14 +193,21 @@ backendStdenv.mkDerivation (
+      # To overcome this, we access the `__spliced` attribute and choose the `hostTarget` attribute.
+      # In the case the `__spliced` attribute doesn't exist, we just use the hook directly (because we're not
+      # cross-compiling).
+      autoAddCudaCompatRunpath.__spliced.hostTarget or autoAddCudaCompatRunpath
```
I was wondering if we should introduce primitives like

```
forPlatform = offsets: p: p.__spliced."pkgs${offsets}" or p;
forBuildHost = forPlatform "BuildHost";
forHostTarget = forPlatform "HostTarget";
# etc
```

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -195,14 +193,21 @@ backendStdenv.mkDerivation (
       # See its documentation in ./setup-hooks/extension.nix.
-      autoAddCudaCompatRunpath
+      # NOTE(@connorbaker): Because autoAddCudaCompatRunpath is in nativeBuildInputs, it tries to use toolchains
+      # from buildPlatform, but that's not what we want. We want to use our host/target toolchains!
```
Hm I need to think about this

SomeoneSerge commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -296,11 +301,14 @@ backendStdenv.mkDerivation (
 
     # For each output, create a symlink to it in the out output.
     # NOTE: We must recreate the out output here, because the setup hook will have deleted it if it was empty.
+    # TODO: Previously we used `meta.getExe lndir` to get the path to lndir, but that doesn't work under
+    # cross-compilation -- whatever machinery Nixpkgs uses to get a version built for hostPlatform (so it can run
+    # during the build) doesn't extend to `meta.getExe`.
```
Too long and informal, but also I imagine that `getExe ((p: p.__spliced.pkgsBuildHost or p) lndir)` would have the same effect. Naturally `nativeBuildInputs` is the preferred solution

ConnorBaker replied:
Resolved in latest; reworded and referenced the tracking issue you made for getting rid of the fat output.

Commit:
cuda-modules/setup-hooks: switch to directory structure

Commit:
cuda-modules: always get *Platform from stdenv

Commit:
cuda-modules/cuda/overrides: simplify callPackage then overrideAttrs pattern

Commit:
cudaPackages.cuda_nvcc: lib must precede static in outputs

Commit:
cuda-modules: add check for duplicate/misordered outputs

Commit:
cuda-modules: update note on use of lndir from path

ConnorBaker commented on pkgs/development/cuda-modules/generic-builders/manifest.nix:
```diff
@@ -195,14 +193,21 @@ backendStdenv.mkDerivation (
       # Check e.g. with `patchelf --print-rpath path/to/my/binary
       autoAddDriverRunpath
       markForCudatoolkitRootHook
+      # To create fat outputs from each component and find a version of `lndir` built for the host platform.
+      lndir
```
I'm not opposed to doing that but I'm a bit swamped at the moment -- I did update the PR to include a link to that issue to track it so it doesn't get lost.

Commit:
cudaPackages.saxpy: Jetson should be supported after CUDA 11.4

Commit:
cuda-modules/cuda/overrides: remove unused callPackage arguments

Commit:
cuda-modules/flags: use cudaAtLeast when possible

Commit:
cuda-modules/flags: ignore platforms in throwIf in isJetsonBuild

Since, even under cross-compilation, we evaluate this flag on multiple platforms, it makes more sense to move the platform check out of the throw condition
and into the boolean return value. The alternative is to restrict all uses of this value to locations which gaurd evaluation so it does not occur when the
host platform is still x86_64.

Commit:
cuda-modules: fix deprecated uses of substituteInPlace replace flag

Commit:
cuda-modules/cuda/overrides: backendStdenv.cc is already part of nativeBuildInputs

Commit:
cuda-modules/cuda/overrides: specify spliced packages for cuda_nvcc postPatch phase

Commit:
cuda-modules/generic-builders/manifest: wip cross-compilation

Commit:
cuda-modules/saxpy: remove CMAKE_VERBOSE_MAKEFILE

Commit:
cuda-modules/setup-hooks/setup-cuda-hook: factor out cc access

Commit:
cuda-modules/setup-hooks/auto-add-cuda-compat-runpath-hook: collapse body like other setup hooks

Commit:
cuda-modules/setup-hooks: wip rewrite and set NIX_DEBUG=1

Commit:
cudaPackages.cuda_nvcc: never has a lib output

Commit:
cuda-modules/setup-hooks: wip

Commit:
cuda-modules/cuda/overrides: cuda_nvcc should not include references to cuda_cudart

Commit:
cudaPackages.nccl: remove unneeded makeFlags, specify splicing, and enable structuredAttrs

Commit:
cuda-modules/generic-builders/manifest: cleanup

Commit:
cuda-modules/setup-hooks/mark-for-cudatoolkit-root-hook: rewrite

Commit:
cudaPackages.saxpy: getDev/getLib would not always select the desired output

Commit:
cuda-modules/setup-hooks/setup-cuda-hook: rewrite

Commit:
cudaPackages.saxpy: attempt manually setting flags for cross

ConnorBaker commented:
Closing as this PR is largely superseded by #301416.
